skada/deep/losses.py

```diff
@@ -189,6 +191,134 @@ def dan_loss(features_s, features_t, sigmas=None):
     return loss
 
 
+def cdd_loss(
+    y_s,
+    features_s,
+    features_t,
+    sigmas=None,
+    distance_threshold=0.5,
+    class_threshold=3,
+):
+    """Define the contrastive domain discrepancy loss based on [33]_.
+
+    Parameters
+    ----------
+    y_s : tensor
+        labels of the source data used to compute the loss.
+    features_s : tensor
+        features of the source data used to compute the loss.
+    features_t : tensor
+        features of the target data used to compute the loss.
+    sigmas : array like, default=None,
+        If array, sigmas used for the multi gaussian kernel.
+        If None, uses sigmas proposed  in [1]_.
+    distance_threshold : float, optional (default=0.5)
+        Distance threshold for discarding the samples that are
+        to far from the centroids.
+    class_threshold : int, optional (default=3)
+        Minimum number of samples in a class to be considered for the loss.
+
+    Returns
+    -------
+    loss : float
+        The loss of the method.
+
+    References
+    ----------
+    .. [33] Kang, G., Jiang, L., Yang, Y., & Hauptmann, A. G. (2019).
+           Contrastive adaptation network for unsupervised domain adaptation.
+           In Proceedings of the IEEE/CVF Conference on Computer Vision
+           and Pattern Recognition (pp. 4893-4902).
+    """
+    n_classes = len(y_s.unique())
+
+    # Calculate source centroids
+    source_centroids = []
+    for c in range(n_classes):
+        mask = y_s == c
+        if mask.sum() > 0:
+            class_features = features_s[mask]
+            normalized_features = F.normalize(class_features, p=2, dim=1)
+            centroid = normalized_features.mean(dim=0)
+            source_centroids.append(centroid)
+
+    source_centroids = torch.stack(source_centroids)
+
+    # Use source centroids to initialize target clustering
+    target_kmeans = SphericalKMeans(
+        n_clusters=n_classes,
+        random_state=0,
+        centroids=source_centroids,
+        device=features_t.device,
+    )
+    target_kmeans.fit(features_t)
+
+    # Predict clusters for target samples
+    cluster_labels_t = target_kmeans.predict(features_t)
+
+    # Discard ambiguous target samples
+    similarities = F.cosine_similarity(
+        features_t.unsqueeze(1), target_kmeans.cluster_centers_.unsqueeze(0)
+    )
+    mask_t = 0.5 * (1 - similarities.max(dim=1)[0]) < distance_threshold
+    features_t = features_t[mask_t]
+    cluster_labels_t = cluster_labels_t[mask_t]
+
+    # Discard ambiguous classes
+    class_counts = torch.bincount(cluster_labels_t, minlength=n_classes)
+    valid_classes = class_counts >= class_threshold
+    mask_t = valid_classes[cluster_labels_t]
+    features_t = features_t[mask_t]
+    cluster_labels_t = cluster_labels_t[mask_t]
+
+    # Define sigmas
```

Comment:
Do you cannot use the mmd distance from DAN?

Reply:
The formula is not exactly the same as for the mmd since before computing each mean we apply a specific mask